src/responder.ts

```diff
@@ -201,77 +204,67 @@ export class EthereumTransactionMiner {
 
     /**
      * Resolves after the transaction receives the first confirmation.
-     * Rejects with `NoNewBlockError` the provider does not receive a new block for `newBlockTimeout` milliseconds.
+     * Rejects with `BlockTimeoutError` if the `this.blockTimeoutDetector` emits a `BLOCK_TIMEOUT_EVENT`.
      * Rejects with `BlockThresholdReachedError` if the transaction is still unconfirmed
      * after `blocksThresholdForStuckTransaction` blocks are mined.
      *
      * @param txHash The transaction hash, returned by `sendTransaction`.
      * @param timeLastBlockReceived Optional; if known, the time when the last block was received by the provider.
-     *                              The provider will be considered unresponsive after `newBlockTimeout` from `timeLastBlockReceived`.
+     *                              The provider will be considered unresponsive after `newBlockTimeout`.
      */
-    public async waitForFirstConfirmation(txHash: string, timeLastBlockReceived: number) {
-        const lastBlockNumberSeen = await this.provider.getBlockNumber();
-
-        // Promise that waits for the first confirmation
-        const firstConfirmationPromise = waitForConfirmations(this.provider, txHash, 1, false);
-
-        // Promise that rejects after WAIT_BLOCKS_BEFORE_RETRYING blocks are mined
-        const firstConfirmationTimeoutPromise = rejectAfterBlocks(
-            this.provider, lastBlockNumberSeen, this.blocksThresholdForStuckTransaction
+    public async waitForFirstConfirmation(txHash: string) {
+        // Promise that waits for the first confirmation, but rejects if still unconfirmed after the threshold
+        const firstConfirmationPromise = this.confirmationObserver.waitForConfirmations(
+            txHash,
+            1,
+            this.blocksThresholdForStuckTransaction,
+            false
         );
 
         // ...but stop with error if no new blocks come for too long
-        const noNewBlockPromise = rejectIfAnyBlockTimesOut(
-            this.provider,
-            timeLastBlockReceived,
-            this.newBlockTimeout,
-            this.pollInterval
-        );
+        const noNewBlockPromise = new Promise((_, reject) => {
+            this.blockTimeoutDetector.once(BlockTimeoutDetector.BLOCK_TIMEOUT_EVENT, () => {
+                reject(new BlockTimeoutError(`No new block received too long; provider unresponsive.`));
+            });
+        });
 
         try {
             // First, wait to get at least 1 confirmation, but throw an error if the transaction is stuck
             // (that is, new blocks are coming, but the transaction is not included)
-            await Promise.race([
-                firstConfirmationPromise,
-                firstConfirmationTimeoutPromise,
-                noNewBlockPromise
-            ]);
+            await Promise.race([firstConfirmationPromise, noNewBlockPromise]);
```

Comment:
Doesnt the noNewBlock promise need to be cancelled?

Reply:
Solved in ffc3e167ad041682416a75ae73c07cb439831e5b.